paderbox/array/sparse.py

```diff
@@ -0,0 +1,701 @@
+import bisect
+import copy
+from dataclasses import dataclass, field
+
+from typing import Tuple, Any
+
+from cached_property import cached_property
+
+from paderbox.array.interval.core import ArrayInterval
+import numpy as np
+import paderbox as pb
+from paderbox.array.interval.util import cy_parse_item, cy_invert_intervals
+import torch
+
+
+def _normalize_shape(shape):
+    if shape is None:
+        # As of now, we only support fixed shapes. This can be changed
+        raise TypeError(f'None shape is not supported')
+
+    if isinstance(shape, int):
+        return shape,
+
+    if not isinstance(shape, (tuple, list)):
+        raise TypeError(f'Invalid shape {shape} of type {type(shape)}')
+
+    if not all(isinstance(s, int) for s in shape):
+        raise TypeError(
+            f'Invalid shape {shape} with elements of type {type(shape[0])}'
+        )
+
+    return tuple(shape)
+
+
+def zeros(shape):
+    return SparseArray(shape=shape)
+
+
+def _check_shape(shape1, shape2):
+    if shape1 != shape2:
+        raise ValueError(f'Shape mismatch: {shape1} {shape2}')
+
+
+def _clip(target, array, onset):
+    offset = onset + array.shape[-1]
+    if onset > target.shape[-1]:
+        return (), ()
+
+    if onset < 0:
+        array = array[..., -onset:]
+        onset = 0
+    if offset > target.shape[-1]:
+        offset = target.shape[-1]
+        array = array[..., :target.shape[-1] - onset]
+    return slice(onset, offset), array
+
+
+# Use dataclass for recursive to_numpy
+@dataclass  # (repr=False, eq=False)
+class _SparseSegment:
+    onset: int
+    array: [np.ndarray, torch.Tensor]
+
+    @property
+    def offset(self):
+        return self.onset + self.array.shape[-1]
+
+    @property
+    def segment_length(self):
+        return self.array.shape[-1]
+
+    @cached_property
+    def is_torch(self):
+        return isinstance(self.array, torch.Tensor)
+
+    def __repr__(self):
+        return (
+            f'{self.__class__.__name__}(onset={self.onset}, '
+            f'length={self.segment_length}: {self.array})'
+        )
+
+
+@dataclass
+class SparseArray:
+    """
+    A sparse array implementation using intervals to define non-zero regions.
+
+    >>> a = SparseArray(shape=(20, ))
+    >>> a
+    SparseArray(shape=(20,))
+    >>> a[5:10] = np.ones(5, dtype=int)
+    >>> a
+    SparseArray(
+      _SparseSegment(onset=5, length=5: [1 1 1 1 1]),
+      shape=(20,))
+    >>> np.asarray(a)
+    array([0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
+    >>> a[15:20] = 2
+    >>> a
+    SparseArray(
+      _SparseSegment(onset=5, length=5: [1 1 1 1 1])
+      _SparseSegment(onset=15, length=5: [2 2 2 2 2]),
+      shape=(20,))
+    >>> np.asarray(a)
+    array([0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 2, 2, 2, 2, 2])
+
+    Overlapping intervals are not allowed
+    >>> a[5:20] = 3
```

Comment:
Can you change this to `a[9:11] = 3` or `a[14:16] = 3`? (i.e. hit the corner case)